src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs

```diff
@@ -607,6 +605,25 @@ private void ValidateIconFile(IEnumerable<IPackageFile> files, string iconPath)
                 // Validate entry
                 var iconPathStripped = PathUtility.StripLeadingDirectorySeparators(iconPath);
 
+                var ext = Path.GetExtension(iconPath);
+
+                if (string.IsNullOrEmpty(ext))
+                {
+                    throw new PackagingException(
+                        NuGetLogCode.NU5044,
+                        string.Format(CultureInfo.CurrentCulture, NuGetResources.IconEmptyExtension, iconPath));
+                }
+
+                if (!string.IsNullOrEmpty(ext) &&
+                        !ext.Equals(".jpeg", StringComparison.OrdinalIgnoreCase) &&
+                        !ext.Equals(".jpg", StringComparison.OrdinalIgnoreCase) &&
+                        !ext.Equals(".png", StringComparison.OrdinalIgnoreCase))
+                {
+                    throw new PackagingException(
+                        NuGetLogCode.NU5045,
+                        string.Format(CultureInfo.CurrentCulture, NuGetResources.IconInvalidExtension, iconPath));
```

Comment:
Do we have documentation somewhere saying that only jpeg and png icons are supported?  If not, should this be a warning instead?  This will be a breaking change for anyone using bmp, gif, maybe other formats that work today.

Reply:
we do: https://github.com/NuGet/Engineering/blob/0d13572de03f947f7d96dca079efd672fdedb5cc/Client.Specs/Embedded-Icons-in-nupkg.md#about-the-icon-image

I should work on completing the spec.

Reply:
But yeah, we already shipped embedded icon without this validation. Warning message looks promising.

Reply:
nuget.org has an image header validation we can rely on: https://github.com/NuGet/NuGetGallery/blob/master/src/NuGetGallery/Extensions/ImageExtensions.cs

Reply:
Ok, I will change to a warning instead of a error.

Reply:
Not sure if it happen for nuget package. But there is concern for user upload script instead of actual image then trigger the script later, we at least need to check this header validation.

Reply:
- We can validate headers once we have the supported extensions (jpeg/jpg, png). We could throw an error when file headers are not JPEG/PNG compliant.
- We could not enforce to accept only those exceptions because it could break existing functionality an other 3rd-party servers.

Reply:
> we do

That's in a private repo, meaning that customers have no way of discovering it, and people implementing their own NuGet servers have no reason to impose the same limitations that nuget.org has for icons. Even if it was public, it's not on docs.microsoft.com/nuget, so I wouldn't consider it product documentation that we can use as justification for making nuget more restrictive than it was previously.

We can talk to PMs to get their input, but I think this should be a warning, not an error (which you already said you'd do).

Reply:
It's on docs: https://github.com/NuGet/docs.microsoft.com-nuget/blob/master/docs/reference/msbuild-targets.md#packing-an-icon-image-file

Reply:
Yeah, I will prototype here the warning to show it to the PMs and get their input.

Reply:
ok, since it is documented publicly, then I'm more ok with it being an error, but I'd still like PM input.

Reply:
Discussed offline. We will show an error when the extension is invalid.

Case | Result
-- | --
No extension | Error: Invalid extension
.jpeg, .jpg, .png | OK
Otherwise | Error: Invalid extension